osutil/disks/gpt.go

```diff
@@ -0,0 +1,146 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package disks
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"os"
+)
+
+type GPTLBA uint64
+type GPTGUID [16]byte
+
+// https://en.wikipedia.org/wiki/GUID_Partition_Table#Partition_table_header_(LBA_1)
+type GPTHeader struct {
+	Signature      [8]byte
+	Revision       uint32
+	HeaderSize     uint32
+	HeaderCRC      uint32
+	Reserved       uint32
+	CurrentLBA     GPTLBA
+	AlternateLBA   GPTLBA
+	FirstUsableLBA GPTLBA
+	LastUsableLBA  GPTLBA
+	DiskGUID       [16]byte
+	EntriesLBA     GPTLBA
+	NEntries       uint32
+	EntrySize      uint32
+	EntriesCRC     uint32
+}
+
+func LoadGPTHeader(devfd *os.File) (GPTHeader, error) {
+	var header GPTHeader
+	rawHeader := make([]byte, 512)
+	read, err := devfd.Read(rawHeader)
+	if err != nil {
+		return GPTHeader{}, err
+	}
+
+	if read < binary.Size(header) {
```

Comment:
I think it would make more sense for the logic to assume it's always either able to read a full sector's worth of data, and if it is not, then it's most likely an error and it should not assume integrity of data is valid? So instead,
```read < len(rawHeader)```

Reply:
Technically, for us, we do not really need the sector to be complete, even though it would be weird. I suppose that could happen with an image file from a network mount. It is just a little more robust to not check against a bigger value, but what we exactly need.